### DISCUSSION TITLE
rounded/filetted corners possible with cadquery?

### DISCUSSION THREAD

**Discussion by pathnirvana** [joshreve/dactyl-keyboard#21] [Ideas]
Many thanks for doing this excellent work. When I was reading about the differences between openscad and cadquery few weeks ago one of the selling points of cadquery was the ability to filet very easily. I wonder if it is possible to add an option to round the corners in the dactyl case.

**Comment by joshreve** [comment]
It is tough because the edges are really erratic.  For my new generator, I use spheroid posts to soften all the edges inherently during construction, but that would really mess up this generator since it was not designed for it.  You can open the step files in FreeCAD and try the fillets or blend and if it seems like it is working I can take a look.  I tried a while ago and just tried it again and most failed because the sliver-like faces are extremely challenging to fillet or blend.  The approach of hulls doesn't really pair well with that.  This is one area production CAD programs are still way ahead, complex fillets, blends, and drafts.

**Reply by pathnirvana** [reply]
I see, makes sense. do you have any case step files checked in somewhere? I see only the plate files in the things directory. Adding filets to the plate step files in fusion 360 is easy - just tried it.

**Reply by joshreve** [reply]
I'm trying to move them to the release files, but have been adding features so I don't have that ready yet.  I can put a specific set in if you are having issues getting the generator running.  

Makes sense that fusion can do it as it has a more advanced engine behind it.  FreeCAD and cadquery have issues when a blend or fillet crosses to a second face, but most production CAD packages are programmed to manage that case.  Unfortunately none of them are free to be an opensource generator.  I've worked with Solidworks, NX, and Creo and they could all make these blends, but few people have access to them.

**Reply by pathnirvana** [reply]
many thanks for the explantion. makes lot of sense.
It would be really helpful if you could checkin some case step files too, specially for folks like me who are not familiar with python. I would be happy to play around with the step files and give you any feedback. 

Also on a side note, have you thought about the ability to generate the sides of the case without using triangle hulls. Maybe shelling or lofting??

**Reply by joshreve** [reply]
I plan on getting them into the releases tonight or in the next couple days.  I know people want an easy way to get the default designs, I'm just mid transition at the moment.  I'll see if I can get it in this afternoon.

Loft was the first thing I tried waaaaaay back to replace the hulls and the problem is the generator is not careful and there are some odd shapes, like two faces you want to connect are actually overlapping.  Hulls just ignore that and create overlapping triangles that union fine, but loft just explodes.  It would really take a new generation methodology to achieve that consistently.  Shelling wouldn't change much from the hulls as it also tends to be craggy.  Essentially the hull command I wrote generates by looking at the points in the shape, finding the hull, and creating a shell to convert to solid.  Only loft does a spline to join and unfortunately it isn't stable enough for some of the shapes.  

In the end I wrote a hull command for cadquery because the other options were less tolerant of questionable geometries.

**Reply by pathnirvana** [reply]
Many thanks @joshreve for the all the good work you are doing.
I too played around with openscad and fusion360 to find any alternate ways to generate the dactyl cases and in all cases it amounts to lot of work.

Also saw your comment on the blended dactyl reddit post. Happy to see lot of interest in this area. looking forward to see the default designs....
